File: frontend/src/types/billing.ts
/**
 * Billing & Usage Quotas Types
 *
 * TypeScript types for subscriptions, billing, invoices, payment methods,
 * usage tracking, and quota management.
 */

// Enums

export enum SubscriptionStatus {
  ACTIVE = 'active',
  TRIALING = 'trialing',
  PAST_DUE = 'past_due',
  CANCELED = 'canceled',
  UNPAID = 'unpaid',
  PAUSED = 'paused',
  INCOMPLETE = 'incomplete',
}

export enum BillingInterval {
  MONTHLY = 'monthly',
  QUARTERLY = 'quarterly',
  YEARLY = 'yearly',
}

export enum PaymentStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  SUCCEEDED = 'succeeded',
  FAILED = 'failed',
  REFUNDED = 'refunded',
  PARTIALLY_REFUNDED = 'partially_refunded',
  CANCELED = 'canceled',
}

export enum InvoiceStatus {
  DRAFT = 'draft',
  OPEN = 'open',
  PAID = 'paid',
  VOID = 'void',
  UNCOLLECTIBLE = 'uncollectible',
}

export enum PaymentMethodType {
  CARD = 'card',
  BANK_TRANSFER = 'bank_transfer',
  PAYPAL = 'paypal',
  INVOICE = 'invoice',
}

export enum UsageMetricType {
  API_CALLS = 'api_calls',
  STORAGE_MB = 'storage_mb',
  QUERY_MINUTES = 'query_minutes',
  DASHBOARD_VIEWS = 'dashboard_views',
  EXPORT_COUNT = 'export_count',
  MEMBER_COUNT = 'member_count',
  CONNECTION_COUNT = 'connection_count',
  SCHEDULED_REPORTS = 'scheduled_reports',
}

export enum CreditType {
  PROMOTIONAL = 'promotional',
  REFERRAL = 'referral',
  COMPENSATION = 'compensation',
  PREPAID = 'prepaid',
}

// Plan Interfaces

export interface PlanFeature {
  code: string;
  name: string;
  description?: string;
  included: boolean;
  limit?: number;
  unit?: string;
}

export interface Plan {
  id: string;
  name: string;
  description?: string;
  tier: number;
  price_monthly: number;
  price_yearly: number;
  currency: string;
  features: PlanFeature[];
  limits: Record<string, number>;
  is_public: boolean;
  is_popular: boolean;
  trial_days: number;
  metadata: Record<string, unknown>;
}

export interface PlanComparison {
  plans: Plan[];
  feature_categories: Array<{
    name: string;
    features: string[];
  }>;
}

// Subscription Interfaces

export interface SubscriptionCreate {
  workspace_id: string;
  plan_id: string;
  billing_interval?: BillingInterval;
  payment_method_id?: string;
  coupon_code?: string;
  trial_days?: number;
}

export interface Subscription {
  id: string;
  workspace_id: string;
  plan_id: string;
  plan_name: string;
  status: SubscriptionStatus;
  billing_interval: BillingInterval;
  current_period_start: string;
  current_period_end: string;
  trial_start?: string;
  trial_end?: string;
  canceled_at?: string;
  cancel_at_period_end: boolean;
  payment_method_id?: string;
  quantity: number;
  unit_amount: number;
  currency: string;
  discount_percent: number;
  coupon_code?: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionUpdate {
  plan_id?: string;
  billing_interval?: BillingInterval;
  payment_method_id?: string;
  quantity?: number;
  coupon_code?: string;
  cancel_at_period_end?: boolean;
}

export interface SubscriptionUsage {
  subscription_id: string;
  period_start: string;
  period_end: string;
  metrics: Record<string, {
    used: number;
    limit: number;
    percent: number;
  }>;
}

export interface SubscriptionPreview {
  current_plan: Plan;
  new_plan: Plan;
  proration_amount: number;
  new_amount: number;
  currency: string;
  effective_date: string;
  credit_applied: number;
}

// Invoice Interfaces

export interface InvoiceLineItem {
  id: string;
  description: string;
  quantity: number;
  unit_amount: number;
  amount: number;
  currency: string;
  period_start?: string;
  period_end?: string;
  metadata: Record<string, unknown>;
}

export interface Invoice {
  id: string;
  number: string;
  workspace_id: string;
  subscription_id?: string;
  status: InvoiceStatus;
  currency: string;
  subtotal: number;
  tax: number;
  tax_percent: number;
  discount: number;
  total: number;
  amount_paid: number;
  amount_due: number;
  line_items: InvoiceLineItem[];
  period_start?: string;
  period_end?: string;
  due_date?: string;
  paid_at?: string;
  pdf_url?: string;
  hosted_invoice_url?: string;
  billing_reason?: string;
  notes?: string;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface InvoiceListResponse {
  invoices: Invoice[];
  total: number;
}

// Payment Interfaces

export interface PaymentMethodCreate {
  type: PaymentMethodType;
  is_default?: boolean;
  card_number?: string;
  card_exp_month?: number;
  card_exp_year?: number;
  card_cvc?: string;
  billing_name?: string;
  billing_email?: string;
  billing_address?: Record<string, string>;
}

export interface PaymentMethod {
  id: string;
  workspace_id: string;
  type: PaymentMethodType;
  is_default: boolean;
  card_brand?: string;
  card_last4?: string;
  card_exp_month?: number;
  card_exp_year?: number;
  billing_name?: string;
  billing_email?: string;
  billing_address?: Record<string, string>;
  created_at: string;
}

export interface Payment {
  id: string;
  workspace_id: string;
  invoice_id?: string;
  subscription_id?: string;
  payment_method_id?: string;
  status: PaymentStatus;
  amount: number;
  currency: string;
  description?: string;
  failure_reason?: string;
  refund_amount: number;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface PaymentListResponse {
  payments: Payment[];
  total: number;
}

// Usage Interfaces

export interface UsageRecord {
  id: string;
  workspace_id: string;
  subscription_id?: string;
  metric_type: UsageMetricType;
  quantity: number;
  timestamp: string;
  metadata: Record<string, unknown>;
}

export interface UsageSummary {
  workspace_id: string;
  period_start: string;
  period_end: string;
  metrics: Record<string, {
    total: number;
    limit: number;
    percent: number;
    daily_breakdown: Record<string, number>;
  }>;
}

export interface UsageAlert {
  id: string;
  workspace_id: string;
  metric_type: UsageMetricType;
  threshold_percent: number;
  notification_channels: string[];
  is_enabled: boolean;
  last_triggered_at?: string;
  created_at: string;
}

export interface UsageAlertCreate {
  metric_type: UsageMetricType;
  threshold_percent: number;
  notification_channels?: string[];
}

// Credit Interfaces

export interface Credit {
  id: string;
  workspace_id: string;
  type: CreditType;
  amount: number;
  currency: string;
  remaining: number;
  description?: string;
  expires_at?: string;
  created_at: string;
}

export interface CreditApplication {
  id: string;
  credit_id: string;
  invoice_id: string;
  amount: number;
  applied_at: string;
}

export interface CreditBalance {
  workspace_id: string;
  total_credits: number;
  available_credits: number;
  expiring_soon: number;
  currency: string;
  credits: Credit[];
}

// Coupon Interfaces

export interface Coupon {
  id: string;
  code: string;
  name: string;
  description?: string;
  discount_type: string;
  discount_value: number;
  currency: string;
  duration: string;
  duration_months?: number;
  max_redemptions?: number;
  redemption_count: number;
  applies_to_plans: string[];
  valid_from?: string;
  valid_until?: string;
  is_active: boolean;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface CouponValidation {
  valid: boolean;
  coupon?: Coupon;
  error_message?: string;
  discount_amount?: number;
}

// Billing Settings Interfaces

export interface BillingSettings {
  workspace_id: string;
  billing_email?: string;
  billing_name?: string;
  billing_address?: Record<string, string>;
  tax_id?: string;
  tax_id_type?: string;
  auto_collection: boolean;
  invoice_prefix?: string;
  default_payment_method_id?: string;
  metadata: Record<string, unknown>;
}

export interface BillingSettingsUpdate {
  billing_email?: string;
  billing_name?: string;
  billing_address?: Record<string, string>;
  tax_id?: string;
  tax_id_type?: string;
  auto_collection?: boolean;
  invoice_prefix?: string;
  default_payment_method_id?: string;
}

// Billing Portal

export interface BillingPortalSession {
  url: string;
  expires_at: string;
}

// Overview Interface

export interface BillingOverview {
  workspace_id: string;
  subscription?: Subscription;
  current_plan?: Plan;
  usage_summary?: UsageSummary;
  credit_balance?: CreditBalance;
  upcoming_invoice?: Invoice;
  payment_methods: PaymentMethod[];
  recent_invoices: Invoice[];
}

// Constants

export const SUBSCRIPTION_STATUS_LABELS: Record<SubscriptionStatus, string> = {
  [SubscriptionStatus.ACTIVE]: 'Active',
  [SubscriptionStatus.TRIALING]: 'Trial',
  [SubscriptionStatus.PAST_DUE]: 'Past Due',
  [SubscriptionStatus.CANCELED]: 'Canceled',
  [SubscriptionStatus.UNPAID]: 'Unpaid',
  [SubscriptionStatus.PAUSED]: 'Paused',
  [SubscriptionStatus.INCOMPLETE]: 'Incomplete',
};

export const BILLING_INTERVAL_LABELS: Record<BillingInterval, string> = {
  [BillingInterval.MONTHLY]: 'Monthly',
  [BillingInterval.QUARTERLY]: 'Quarterly',
  [BillingInterval.YEARLY]: 'Yearly',
};

export const PAYMENT_STATUS_LABELS: Record<PaymentStatus, string> = {
  [PaymentStatus.PENDING]: 'Pending',
  [PaymentStatus.PROCESSING]: 'Processing',
  [PaymentStatus.SUCCEEDED]: 'Succeeded',
  [PaymentStatus.FAILED]: 'Failed',
  [PaymentStatus.REFUNDED]: 'Refunded',
  [PaymentStatus.PARTIALLY_REFUNDED]: 'Partially Refunded',
  [PaymentStatus.CANCELED]: 'Canceled',
};

export const INVOICE_STATUS_LABELS: Record<InvoiceStatus, string> = {
  [InvoiceStatus.DRAFT]: 'Draft',
  [InvoiceStatus.OPEN]: 'Open',
  [InvoiceStatus.PAID]: 'Paid',
  [InvoiceStatus.VOID]: 'Void',
  [InvoiceStatus.UNCOLLECTIBLE]: 'Uncollectible',
};

export const PAYMENT_METHOD_LABELS: Record<PaymentMethodType, string> = {
  [PaymentMethodType.CARD]: 'Credit Card',
  [PaymentMethodType.BANK_TRANSFER]: 'Bank Transfer',
  [PaymentMethodType.PAYPAL]: 'PayPal',
  [PaymentMethodType.INVOICE]: 'Invoice',
};

export const USAGE_METRIC_LABELS: Record<UsageMetricType, string> = {
  [UsageMetricType.API_CALLS]: 'API Calls',
  [UsageMetricType.STORAGE_MB]: 'Storage (MB)',
  [UsageMetricType.QUERY_MINUTES]: 'Query Minutes',
  [UsageMetricType.DASHBOARD_VIEWS]: 'Dashboard Views',
  [UsageMetricType.EXPORT_COUNT]: 'Exports',
  [UsageMetricType.MEMBER_COUNT]: 'Team Members',
  [UsageMetricType.CONNECTION_COUNT]: 'Connections',
  [UsageMetricType.SCHEDULED_REPORTS]: 'Scheduled Reports',
};

export const CREDIT_TYPE_LABELS: Record<CreditType, string> = {
  [CreditType.PROMOTIONAL]: 'Promotional',
  [CreditType.REFERRAL]: 'Referral',
  [CreditType.COMPENSATION]: 'Compensation',
  [CreditType.PREPAID]: 'Prepaid',
};

// Helper Functions

export function formatCurrency(amount: number, currency: string = 'USD'): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency,
  }).format(amount);
}

export function formatPrice(plan: Plan, interval: BillingInterval): string {
  const price = interval === BillingInterval.YEARLY ? plan.price_yearly : plan.price_monthly;
  if (price === 0 && plan.tier === 0) return 'Free';
  if (price === 0 && plan.tier > 3) return 'Contact Us';
  return formatCurrency(price);
}

export function getYearlySavings(plan: Plan): number {
  const monthly_total = plan.price_monthly * 12;
  const yearly_price = plan.price_yearly;
  return monthly_total - yearly_price;
}

export function getYearlySavingsPercent(plan: Plan): number {
  const savings = getYearlySavings(plan);
  const monthly_total = plan.price_monthly * 12;
  if (monthly_total === 0) return 0;
  return Math.round((savings / monthly_total) * 100);
}

export function getSubscriptionStatusColor(status: SubscriptionStatus): string {
  const colors: Record<SubscriptionStatus, string> = {
    [SubscriptionStatus.ACTIVE]: 'green',
    [SubscriptionStatus.TRIALING]: 'blue',
    [SubscriptionStatus.PAST_DUE]: 'orange',
    [SubscriptionStatus.CANCELED]: 'gray',
    [SubscriptionStatus.UNPAID]: 'red',
    [SubscriptionStatus.PAUSED]: 'yellow',
    [SubscriptionStatus.INCOMPLETE]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getInvoiceStatusColor(status: InvoiceStatus): string {
  const colors: Record<InvoiceStatus, string> = {
    [InvoiceStatus.DRAFT]: 'gray',
    [InvoiceStatus.OPEN]: 'blue',
    [InvoiceStatus.PAID]: 'green',
    [InvoiceStatus.VOID]: 'gray',
    [InvoiceStatus.UNCOLLECTIBLE]: 'red',
  };
  return colors[status] || 'gray';
}

export function getPaymentStatusColor(status: PaymentStatus): string {
  const colors: Record<PaymentStatus, string> = {
    [PaymentStatus.PENDING]: 'yellow',
    [PaymentStatus.PROCESSING]: 'blue',
    [PaymentStatus.SUCCEEDED]: 'green',
    [PaymentStatus.FAILED]: 'red',
    [PaymentStatus.REFUNDED]: 'gray',
    [PaymentStatus.PARTIALLY_REFUNDED]: 'orange',
    [PaymentStatus.CANCELED]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getUsagePercent(used: number, limit: number): number {
  if (limit === -1) return 0; // Unlimited
  if (limit === 0) return used > 0 ? 100 : 0;
  return Math.min(100, Math.round((used / limit) * 100));
}

export function getUsageStatus(percent: number): 'ok' | 'warning' | 'critical' {
  if (percent >= 100) return 'critical';
  if (percent >= 80) return 'warning';
  return 'ok';
}

export function formatLimit(limit: number, unit?: string): string {
  if (limit === -1) return 'Unlimited';
  if (unit === 'GB') return `${limit / 1000} GB`;
  return limit.toLocaleString() + (unit ? ` ${unit}` : '');
}

export function getDaysUntilRenewal(subscription: Subscription): number {
  const end = new Date(subscription.current_period_end);
  const now = new Date();
  const diff = end.getTime() - now.getTime();
  return Math.max(0, Math.ceil(diff / (1000 * 60 * 60 * 24)));
}

export function isTrialEnding(subscription: Subscription, daysThreshold: number = 3): boolean {
  if (subscription.status !== SubscriptionStatus.TRIALING || !subscription.trial_end) {
    return false;
  }
  const end = new Date(subscription.trial_end);
  const now = new Date();
  const diff = end.getTime() - now.getTime();
  const days = Math.ceil(diff / (1000 * 60 * 60 * 24));
  return days <= daysThreshold;
}

export function getCardBrandIcon(brand?: string): string {
  const icons: Record<string, string> = {
    visa: '💳',
    mastercard: '💳',
    amex: '💳',
    discover: '💳',
  };
  return icons[brand?.toLowerCase() || ''] || '💳';
}

File: backend/app/schemas/billing.py
"""
Billing & Usage Quotas Schemas

Pydantic schemas for subscriptions, billing, invoices, payment methods,
usage tracking, and quota management.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class SubscriptionStatus(str, Enum):
    """Subscription status"""
    ACTIVE = "active"
    TRIALING = "trialing"
    PAST_DUE = "past_due"
    CANCELED = "canceled"
    UNPAID = "unpaid"
    PAUSED = "paused"
    INCOMPLETE = "incomplete"


class BillingInterval(str, Enum):
    """Billing interval"""
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    YEARLY = "yearly"


class PaymentStatus(str, Enum):
    """Payment status"""
    PENDING = "pending"
    PROCESSING = "processing"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    REFUNDED = "refunded"
    PARTIALLY_REFUNDED = "partially_refunded"
    CANCELED = "canceled"


class InvoiceStatus(str, Enum):
    """Invoice status"""
    DRAFT = "draft"
    OPEN = "open"
    PAID = "paid"
    VOID = "void"
    UNCOLLECTIBLE = "uncollectible"


class PaymentMethodType(str, Enum):
    """Payment method types"""
    CARD = "card"
    BANK_TRANSFER = "bank_transfer"
    PAYPAL = "paypal"
    INVOICE = "invoice"


class UsageMetricType(str, Enum):
    """Types of usage metrics"""
    API_CALLS = "api_calls"
    STORAGE_MB = "storage_mb"
    QUERY_MINUTES = "query_minutes"
    DASHBOARD_VIEWS = "dashboard_views"
    EXPORT_COUNT = "export_count"
    MEMBER_COUNT = "member_count"
    CONNECTION_COUNT = "connection_count"
    SCHEDULED_REPORTS = "scheduled_reports"


class CreditType(str, Enum):
    """Types of credits"""
    PROMOTIONAL = "promotional"
    REFERRAL = "referral"
    COMPENSATION = "compensation"
    PREPAID = "prepaid"


# Plan Models

class PlanFeature(BaseModel):
    """Feature included in a plan"""
    code: str
    name: str
    description: Optional[str] = None
    included: bool = True
    limit: Optional[int] = None  # None = unlimited
    unit: Optional[str] = None


class Plan(BaseModel):
    """Subscription plan"""
    id: str
    name: str
    description: Optional[str] = None
    tier: int  # 0=free, 1=starter, 2=pro, 3=business, 4=enterprise
    price_monthly: float
    price_yearly: float
    currency: str = "USD"
    features: list[PlanFeature] = Field(default_factory=list)
    limits: dict[str, int] = Field(default_factory=dict)
    is_public: bool = True
    is_popular: bool = False
    trial_days: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class PlanComparison(BaseModel):
    """Plan comparison for pricing page"""
    plans: list[Plan]
    feature_categories: list[dict[str, Any]]


# Subscription Models

class SubscriptionCreate(BaseModel):
    """Create subscription"""
    workspace_id: str
    plan_id: str
    billing_interval: BillingInterval = BillingInterval.MONTHLY
    payment_method_id: Optional[str] = None
    coupon_code: Optional[str] = None
    trial_days: Optional[int] = None


class Subscription(BaseModel):
    """Subscription details"""
    id: str
    workspace_id: str
    plan_id: str
    plan_name: str
    status: SubscriptionStatus
    billing_interval: BillingInterval
    current_period_start: datetime
    current_period_end: datetime
    trial_start: Optional[datetime] = None
    trial_end: Optional[datetime] = None
    canceled_at: Optional[datetime] = None
    cancel_at_period_end: bool = False
    payment_method_id: Optional[str] = None
    quantity: int = 1
    unit_amount: float
    currency: str = "USD"
    discount_percent: float = 0
    coupon_code: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class SubscriptionUpdate(BaseModel):
    """Update subscription"""
    plan_id: Optional[str] = None
    billing_interval: Optional[BillingInterval] = None
    payment_method_id: Optional[str] = None
    quantity: Optional[int] = None
    coupon_code: Optional[str] = None
    cancel_at_period_end: Optional[bool] = None


class SubscriptionUsage(BaseModel):
    """Subscription usage summary"""
    subscription_id: str
    period_start: datetime
    period_end: datetime
    metrics: dict[str, dict[str, Any]]  # metric -> {used, limit, percent}


# Invoice Models

class InvoiceLineItem(BaseModel):
    """Invoice line item"""
    id: str
    description: str
    quantity: int
    unit_amount: float
    amount: float
    currency: str
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class Invoice(BaseModel):
    """Invoice"""
    id: str
    number: str
    workspace_id: str
    subscription_id: Optional[str] = None
    status: InvoiceStatus
    currency: str = "USD"
    subtotal: float
    tax: float = 0
    tax_percent: float = 0
    discount: float = 0
    total: float
    amount_paid: float = 0
    amount_due: float
    line_items: list[InvoiceLineItem] = Field(default_factory=list)
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    due_date: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    pdf_url: Optional[str] = None
    hosted_invoice_url: Optional[str] = None
    billing_reason: Optional[str] = None
    notes: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class InvoiceListResponse(BaseModel):
    """List invoices response"""
    invoices: list[Invoice]
    total: int


# Payment Models

class PaymentMethodCreate(BaseModel):
    """Create payment method"""
    type: PaymentMethodType
    is_default: bool = False
    # Card details (if type=card)
    card_number: Optional[str] = None
    card_exp_month: Optional[int] = None
    card_exp_year: Optional[int] = None
    card_cvc: Optional[str] = None
    # Billing details
    billing_name: Optional[str] = None
    billing_email: Optional[str] = None
    billing_address: Optional[dict[str, str]] = None


class PaymentMethod(BaseModel):
    """Payment method"""
    id: str
    workspace_id: str
    type: PaymentMethodType
    is_default: bool
    # Card info (masked)
    card_brand: Optional[str] = None
    card_last4: Optional[str] = None
    card_exp_month: Optional[int] = None
    card_exp_year: Optional[int] = None
    # Billing details
    billing_name: Optional[str] = None
    billing_email: Optional[str] = None
    billing_address: Optional[dict[str, str]] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class Payment(BaseModel):
    """Payment record"""
    id: str
    workspace_id: str
    invoice_id: Optional[str] = None
    subscription_id: Optional[str] = None
    payment_method_id: Optional[str] = None
    status: PaymentStatus
    amount: float
    currency: str = "USD"
    description: Optional[str] = None
    failure_reason: Optional[str] = None
    refund_amount: float = 0
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class PaymentListResponse(BaseModel):
    """List payments response"""
    payments: list[Payment]
    total: int


# Usage Models

class UsageRecord(BaseModel):
    """Usage record"""
    id: str
    workspace_id: str
    subscription_id: Optional[str] = None
    metric_type: UsageMetricType
    quantity: float
    timestamp: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)


class UsageSummary(BaseModel):
    """Usage summary for a period"""
    workspace_id: str
    period_start: datetime
    period_end: datetime
    metrics: dict[str, dict[str, Any]]  # metric -> {total, limit, percent, daily_breakdown}


class UsageAlert(BaseModel):
    """Usage alert"""
    id: str
    workspace_id: str
    metric_type: UsageMetricType
    threshold_percent: int  # e.g., 80 for 80%
    notification_channels: list[str] = Field(default_factory=list)
    is_enabled: bool = True
    last_triggered_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class UsageAlertCreate(BaseModel):
    """Create usage alert"""
    metric_type: UsageMetricType
    threshold_percent: int
    notification_channels: list[str] = Field(default_factory=list)


# Credit Models

class Credit(BaseModel):
    """Account credit"""
    id: str
    workspace_id: str
    type: CreditType
    amount: float
    currency: str = "USD"
    remaining: float
    description: Optional[str] = None
    expires_at: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class CreditApplication(BaseModel):
    """Credit application to invoice"""
    id: str
    credit_id: str
    invoice_id: str
    amount: float
    applied_at: datetime = Field(default_factory=datetime.utcnow)


class CreditBalance(BaseModel):
    """Credit balance summary"""
    workspace_id: str
    total_credits: float
    available_credits: float
    expiring_soon: float
    currency: str = "USD"
    credits: list[Credit]


# Coupon Models

class Coupon(BaseModel):
    """Discount coupon"""
    id: str
    code: str
    name: str
    description: Optional[str] = None
    discount_type: str  # "percent" or "fixed"
    discount_value: float
    currency: str = "USD"
    duration: str  # "once", "forever", "repeating"
    duration_months: Optional[int] = None
    max_redemptions: Optional[int] = None
    redemption_count: int = 0
    applies_to_plans: list[str] = Field(default_factory=list)
    valid_from: Optional[datetime] = None
    valid_until: Optional[datetime] = None
    is_active: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class CouponValidation(BaseModel):
    """Coupon validation result"""
    valid: bool
    coupon: Optional[Coupon] = None
    error_message: Optional[str] = None
    discount_amount: Optional[float] = None


# Billing Settings

class BillingSettings(BaseModel):
    """Workspace billing settings"""
    workspace_id: str
    billing_email: Optional[str] = None
    billing_name: Optional[str] = None
    billing_address: Optional[dict[str, str]] = None
    tax_id: Optional[str] = None
    tax_id_type: Optional[str] = None
    auto_collection: bool = True
    invoice_prefix: Optional[str] = None
    default_payment_method_id: Optional[str] = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class BillingSettingsUpdate(BaseModel):
    """Update billing settings"""
    billing_email: Optional[str] = None
    billing_name: Optional[str] = None
    billing_address: Optional[dict[str, str]] = None
    tax_id: Optional[str] = None
    tax_id_type: Optional[str] = None
    auto_collection: Optional[bool] = None
    invoice_prefix: Optional[str] = None
    default_payment_method_id: Optional[str] = None


# Billing Portal

class BillingPortalSession(BaseModel):
    """Billing portal session"""
    url: str
    expires_at: datetime


# Response Models

class BillingOverview(BaseModel):
    """Billing overview for dashboard"""
    workspace_id: str
    subscription: Optional[Subscription] = None
    current_plan: Optional[Plan] = None
    usage_summary: Optional[UsageSummary] = None
    credit_balance: Optional[CreditBalance] = None
    upcoming_invoice: Optional[Invoice] = None
    payment_methods: list[PaymentMethod] = Field(default_factory=list)
    recent_invoices: list[Invoice] = Field(default_factory=list)


class SubscriptionPreview(BaseModel):
    """Preview subscription change"""
    current_plan: Plan
    new_plan: Plan
    proration_amount: float
    new_amount: float
    currency: str
    effective_date: datetime
    credit_applied: float = 0


# Constants

PLAN_FEATURES = {
    "dashboards": {"name": "Dashboards", "unit": "dashboards"},
    "charts": {"name": "Charts", "unit": "charts"},
    "connections": {"name": "Data Connections", "unit": "connections"},
    "members": {"name": "Team Members", "unit": "members"},
    "storage": {"name": "Storage", "unit": "GB"},
    "api_calls": {"name": "API Calls", "unit": "calls/month"},
    "query_minutes": {"name": "Query Execution", "unit": "minutes/month"},
    "scheduled_reports": {"name": "Scheduled Reports", "unit": "reports"},
    "alerts": {"name": "Alerts", "unit": "alerts"},
    "ai_queries": {"name": "AI Queries", "unit": "queries/month"},
    "support": {"name": "Support", "unit": ""},
    "sso": {"name": "SSO/SAML", "unit": ""},
    "audit_logs": {"name": "Audit Logs", "unit": "days retention"},
    "custom_branding": {"name": "Custom Branding", "unit": ""},
    "api_access": {"name": "API Access", "unit": ""},
}

DEFAULT_PLANS = [
    {
        "id": "plan_free",
        "name": "Free",
        "tier": 0,
        "price_monthly": 0,
        "price_yearly": 0,
        "limits": {
            "dashboards": 5,
            "charts": 20,
            "connections": 2,
            "members": 3,
            "storage": 100,
            "api_calls": 100,
            "query_minutes": 10,
            "scheduled_reports": 0,
            "alerts": 0,
        },
    },
    {
        "id": "plan_starter",
        "name": "Starter",
        "tier": 1,
        "price_monthly": 29,
        "price_yearly": 290,
        "limits": {
            "dashboards": 20,
            "charts": 100,
            "connections": 5,
            "members": 5,
            "storage": 500,
            "api_calls": 1000,
            "query_minutes": 60,
            "scheduled_reports": 5,
            "alerts": 10,
        },
    },
    {
        "id": "plan_pro",
        "name": "Pro",
        "tier": 2,
        "price_monthly": 79,
        "price_yearly": 790,
        "is_popular": True,
        "limits": {
            "dashboards": 100,
            "charts": 500,
            "connections": 20,
            "members": 25,
            "storage": 5000,
            "api_calls": 10000,
            "query_minutes": 300,
            "scheduled_reports": 25,
            "alerts": 50,
        },
    },
    {
        "id": "plan_business",
        "name": "Business",
        "tier": 3,
        "price_monthly": 199,
        "price_yearly": 1990,
        "limits": {
            "dashboards": 500,
            "charts": 2500,
            "connections": 50,
            "members": 100,
            "storage": 25000,
            "api_calls": 100000,
            "query_minutes": 1000,
            "scheduled_reports": 100,
            "alerts": 200,
        },
    },
    {
        "id": "plan_enterprise",
        "name": "Enterprise",
        "tier": 4,
        "price_monthly": 0,  # Custom pricing
        "price_yearly": 0,
        "limits": {
            "dashboards": -1,  # Unlimited
            "charts": -1,
            "connections": -1,
            "members": -1,
            "storage": -1,
            "api_calls": -1,
            "query_minutes": -1,
            "scheduled_reports": -1,
            "alerts": -1,
        },
    },
]


# Helper Functions

def calculate_proration(
    current_plan: Plan,
    new_plan: Plan,
    days_remaining: int,
    total_days: int,
    billing_interval: BillingInterval,
) -> float:
    """Calculate proration amount for plan change."""
    current_price = current_plan.price_yearly if billing_interval == BillingInterval.YEARLY else current_plan.price_monthly
    new_price = new_plan.price_yearly if billing_interval == BillingInterval.YEARLY else new_plan.price_monthly

    if billing_interval == BillingInterval.YEARLY:
        daily_rate_current = current_price / 365
        daily_rate_new = new_price / 365
    else:
        daily_rate_current = current_price / 30
        daily_rate_new = new_price / 30

    # Credit for unused time on current plan
    credit = daily_rate_current * days_remaining
    # Charge for new plan's remaining time
    charge = daily_rate_new * days_remaining

    return round(charge - credit, 2)


def is_upgrade(current_plan: Plan, new_plan: Plan) -> bool:
    """Check if plan change is an upgrade."""
    return new_plan.tier > current_plan.tier


def get_usage_percent(used: float, limit: int) -> float:
    """Calculate usage percentage."""
    if limit == -1:  # Unlimited
        return 0
    if limit == 0:
        return 100 if used > 0 else 0
    return min(100, round((used / limit) * 100, 2))
